File: 2 - Regression/3 - Polynomial Linear Regression/polynomial_regression.py
# -*- coding: utf-8 -*-
#import all the required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#data preprocessing
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

#dont split the data, minimal data available

#no feature scaling as well

#simple linear regression
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,y)

from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)

lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly ,y)

#visualistaions for linear regression model
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Linear Regression')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

#visualisations for polynomial regression model
X_grid = np.arange(max(X), min(X), 0.1)
X_grid = X_grid.reshape((len(X_grid),1))
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg2.predict(poly_reg.fit_transform(X)), color = 'blue')
plt.title('Polynomial Regression')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()


#predicting a new result with Linear Regression
lin_reg.predict(6.5)

#predicting the new result with Polynomail Linear Regression
lin_reg2.predict(poly_reg.fit_transform(6.5))

File: 2 - Regression/3 - Polynomial Linear Regression/polynomial_regression.R
#dataset preprocessing

dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

#splitting of dataset into training and test data is not required

#feature scaling is also not required

#fitting linear regression model
lin_reg = lm(formula = Salary ~ .,
             data = dataset)

#fitting polynomial regression model
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4

poly_reg = lm(formula = Salary ~ .,
              data = dataset)

#adding visualisations for linear regression model
library(ggplot2)
ggplot()+
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red')+
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            color = 'blue')+
  ggtitle('Linear Regression')+
  xlab('Level')+
  ylab('Salary')


x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot()+
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red')+
  geom_line(aes(x = x_grid, y = predict(poly_reg,
                                        newdata = data.frame(Level = x_grid,
                                                             Level2 = x_grid^2,
                                                             Level3 = x_grid^3,
                                                             Level4 = x_grid^4))),
            color = 'blue')+
  ggtitle('Polynomial Regression')+
  xlab('Level')+
  ylab('Salary')

#predicting a new result with linear regression model
y_pred = predict(lin_reg, data.frame(Level = 6.5))

#predicting the new result using the polynomial regression model
y_pred = predict(poly_reg, data.frame(Level = 6.5,
                                      Level2 = 6.5^2,
                                      Level3 = 6.5^3,
                                      Level4 = 6.5^4))

